### ISSUE TITLE
pnumpy: request ufunc like hooks for conversion functions, reduce funcs, better ufunc hooking with callback

### ISSUE THREAD

**Issue by tdimitri** [numpy/numpy#18028]
## Feature

These are 5 groups of feature hook requests from the [pnumpy](https://github.com/Quansight/numpy-threading-extensions) project.  @mattip request I file issues here with attention to @seberg.
We desire to make numpy work even better by adding pnumpy which can distribute work across large arrays to different CPU cores.  Due to the nature of CPU L1/L2/L3 caching, especially on the AMD which uses [MOESI](https://en.wikipedia.org/wiki/MOESI_protocol), it is beneficial to hook **all popular array loops**.  If the routines are not hooked, the "pnumpy" project will perform poorly only because we lack the hooks from the parent project "numpy", even though the code and concepts work well.

**Hook 1**: conversion functions
When an array is converted from one dtype to another, such as int32 to float64, we would like to hook this loop.  This is necessary to speed operations like np.true_divide(a,3)  where a is int32.  Although we can hook "true_divide" there is a behind the scenes conversion of the int32 array (in this example) which is single threaded, and destroys the cache on another core for large arrays.

**Hook 2**: reduce functions
Although we can hook functions like np.sum, setbufsize gets in the way.  pnumpy's low level loops can internally upcast float32 to float64 and does not require setbufsize upcasting or additional conversions.  With this change, a parallel sum will run 10x faster on a large array.  Further functions like nansum, argmax, nanargmax, var, std, and others are missing ufunc style hooks.  For example "std" can be sped up 20x for large arrays with a hook in place.

**Hook 3**: array initialization functions
Because of how pnumpy works with caches across cores, it is benefical to hook functions which are often used to initalize arrays.  Common functions include zeros, ones, arange, and random number generation.

**Hook 4**: how to hook a call signature that does not yet exist?
For pnumpy we have some additional loops that can do things like: compare int64 to uint64 without upcasting.  In numpy today, such a comparison results in upcasting to float64, with mantissa loss and thus an imperfect comparison.  We would like to via PyUFunc_ReplaceLoopBySignature the "compare_" functions with signatures like (int64, uint64, bool) but I am not sure how.
I am also not sure how to hook true_divide with a signature like (int32, int32, float64).  Really I prefer (int32, any, float64) and (any, int32, float64) because inside the true_divide loop, a conversion can be passed as a parameter.  Inside the true_divide loop, it will convert int32-> float64 on the stack, such that it fits in the L1 cache.  Then it will internally upcast the other array (i.e., int32) and divide with the converted result still on the stack, and lastly place the result in the output array.  This will make true_divide run much faster when dividing integers.

**Hook 5**: allow for user callback arg
There currently is no user arg callback in the ufunc routine.  Because of this, thousands of stubs have to be written to feed back into to a common function.  See [stubs.h](https://github.com/Quansight/numpy-threading-extensions/blob/main/src/pnumpy/stubs.h).  This entire file goes away if we can pass a "void *" callback argument in the function PyUFunc_ReplaceLoopBySignature.

Thank you

**Comment by seberg** [comment]
Maybe you can schedule a meeting for specific points, to be honest, right now I can't do much more than point to my [NEP 43](https://numpy.org/neps/nep-0043-extensible-ufuncs.html) and its `ArrayMethod`, I am waiting for my PR which rewrites casting in terms of this `ArrayMethod` in gh-17863, you may have a look. (This is not the final say there, the signatures still need fixing at the very least.)

Going through your points:

1. You really want to replace a cast entirely.  After the PR I linked above, these will be defined by ArrayMethod's (internally). We could add API to swap out the actual `ArrayMethod` used.

2. Specializing reduce is not on my horizon, it simply has no priority for me personally. I feel there are multiple points here:
   * The "buffering" issue, this is gh-17471 (although last time I looked I did not see much speed improvement). It may be tricky, but in the end a fairly separate issue not requiring a new hook.
   * Specialized loops for a higher output dtype are part of Point 4
   * Using a larger type internally may require a more proper `teardown` than I currently planned for in NEP 43, although maybe it doesn't matter much.  For what you do, hooking one level higher makes more sense anyway probably?
   * There are some reduce-like functions (`std`), which would be nice if they were `gufunc`'s.  For some of them the only issue may be that ufuncs do not understand keyword arguments. Maybe some should be be ported but are not yet. (I have no plan for kwarg support, it could be nice, but I am not sure how it should look like.) 

3. This sounds mostly like the array-allocator PR that Matti opened recently.  For some functions, implementing them as a `gufunc` would be the nicest solution again. Although I am not sure how feasible that is.

4. I have plans for that, it is specifically mentioned in NEP 43 (although in a very side point probably). You could probably do it in current NumPy if you were to insert types into the loops (instead of adding them at the end), I am not sure that can be done (without messing with our ABI). So don't do it, because not breaking parts of that ABI when implementing NEP 43 will just cause pain for everyone.

5. Again, see NEP 43 and we can discuss how expanding this will help you. Right now there is a lot of "yeah, we probably want to make this public, but there is this argument `move_references` that is annoying, so I am not sure".


So, half of the points come down to: Please have a look at NEP 43, and lets discuss it!  We can also go through the NEP together.  I have to admit, that ufuncs are not the next step on my agenda right now (there are just other things to bang my head against first), and NEP 43 is not finalized and will require more public API to serve your purpose.

But, I am very sure about the `ArrayMethod` scheme even if some details are tricky or fuzzy. And making sure that it can suit your needs will be very valuable!  Other than that, it is slow progress for me, but I am sure we will get there eventually, although help to make it sooner is appreciated; and I have to stress that figuring out final public API relevant for you is a backburner for me, because I am focused on new DTypes and fixing public API is painful, so I have to avoid that where I can (for which I mean, it would be much easier if it was proposed by someone else).

**Comment by tdimitri** [comment]
Ok I will schedule a meeting.  You may join a meeting this Tuesday at 10am EST if you like where we go over this.

In general, the data analysis community benefits from hooking common numpy loops because an external loop may: 1) use parallelism, 2) have a better written loop (likely with vector intrinsics) and 3) future dist. computing

I have read NEP 43 and I hope it addresses the user call back arg.  

However, I do not believe NEP 43 addressed the central issue which is:
I believe external packages should be able to hook common np routines like np.sum, np. std, nanarg, putmask, where, etc..  
The main role of numpy is to parse and pre-flight check the parameters.  Then once everything is ready (for instance lists converted to arrays, dtypes checked, lengths checked) there should be a function through a call table, which has a predefined call signature and is hookable.

That is really it.  I am more than happy to take existing numpy code, modify it and show that many routines have two parts to it -- step 1, processing the python inputs and making sure the function can happen -- step 2, calling that function with a predefined call signature.

Until then, requests like this will keep rolling in.  I believe we should believe taking numpy function and dividing them into step 1) where we process the inputs, pre-flight check and then step 2) where we call through a call table with a predefined API/signature.

I am more than write the code demonstrating one way -- although everyone here seems more than skilled enough themselves.

**Comment by seberg** [comment]
Happy to talk on Tuesday at that time.  I had written some more, but trying to just keep it brief.

I am realizing now, that you have a different angle (I think you mentioned that in the array allocator discussion), and aim for a much deeper hook.  Some (random notes):

* I find this interesting, because we  have little prior art and I do not know it. Previous suggestions were usually a "replace everything" approach (i.e. on the Python level), which ends up being comparable to monkey-patching.
* We have to juggle multiple types of "dispatching":
  1. Dispatching for DTypes (which can be user defined)
  2. Dispatching of array-objects through `__array_function__` and `__array_ufunc__`
  3. Dispatching to "implementations" for core algorithms in NumPy (what you are suggesting).
  And I am not quite sure yet how we can juggle the DType and your low-level dispatching.  `ArrayMethod` was my proposed solution to the DType dispatching problem. I *hope* we can marry that with the last one (by replacing method or by hooking into them, I do not know).  But you are right, I have no clear idea on that.
* What is your view on API changes in NumPy? Say we create such a table, would it be perfectly OK if we simply stop using one of the entries because we want to replace it with a new API?
* Maybe we are skilled enough, but I do not think we any have experience, and with this type of thing that will go a *very long* way to getting it right.

**Comment by tdimitri** [comment]
When I read the above comments, I sense a strong desire to improve numpy.  The motivation is there, and that is wonderful -- even after possible scars from bugs/feature requests.  That is refreshing.

Other thoughts I have:  Are the changes worth the effort?  Will it significantly improve numpy? Do we know how to write the API so that it lasts and handles future changes?

From my perspective, I believe it will significantly improve numpy for larger arrays because I have done this before (older riptable/riptide_cpp) [project](https://github.com/rtosholdings) and I am aware of benchmarks and which tools the quants (at the company I work for) choose to use.  For me, this is the second version of this code.  I am learning from my mistakes in riptable project and writing it better in the pnumpy project.

For me (since I have already convinced myself about parallel loops), the next question is "how do we do this so it lasts?"

numpy has already evolved into the de facto array math package for python. Now I believe we can make it even better by allowing other languages to benefit as well (for instance rust, or java, or C#).  Now numpy-like APIs are truly cross-language.

Once numpy is bifurcated into two parts (and this can take years to happen and that is ok)... I believe additional benefits will come.

The TWO PARTS ARE:
[numpy python - front end parsing, pre-flight check, etc] --->  [ portable C lib, template based loop routines, parallel aware, alloc]
There is figurative Chinese wall between the two parts which allows us decouple and independently work on them.
Someone could write rust code and hook into the second part - the portable C lib, and benefit from all the existing work.

Existing ufuncs are a good example of an independent call interface - it just passes memory pointers, input arrays, output arrays, lengths and strides - and has nothing to do with Python.

So imagine we slowly tease out all the loops in the numpy (could take years and have many releases -- each release moves more code into the portable 'C' lib layer).  Now we have decoupled python from the back end loops.

Once we accomplish this we can
1) allocate memory on the GPU and run all the loops on GPU memory.  the user code does not change at all
2) use multiple cores on the CPU to divide up work for large arrays
3) for even more massive arrays, use distributed computing, and each of those computers may then use individual CPU cores, or even GPUs (mix and match)

The user does not change their code (not one line) because we have hooked BELOW numpy instead of ABOVE.  This is better than the DASK design which sits on top and requires code changes.

I believe that is the overall goal - to split numpy into those two parts.  I am sure this requires more thought and discussion.  I know it will take time and work to surgically go in and tease out all the loops, memory allocators, conversions, etc.  I know it will come in pieces.  That is all ok.  I am hoping we can agree that is general direction to take numpy in the future.

If we can get rough agreement to that direction, then we can have more discussions on good API design.  I have worked on 'C' interface specs before -- worked on the NDIS (Network Driver Interface Spec) and WANDIS specs a long time ago.  Network interface specs also involve memory allocation, multiple cores, protocol stacks, and even more complex problems like connections coming up and down (cancel, timeout).

I apologize for the long winded comments, but I think it is important that we understand each other's motivations, skills. and points of view.  I will email Matti about the 10AM EST phone call on Tuesday, Dec 22.

**Comment by seberg** [comment]
I have not made up my mind of course.  The discussion about the allocator made sense to me and in part I am interested because I have hope that the long-term maintenance and API burden could be very small and maybe even negative (while allowing NumPy to gain momentum for internal improvements).

That said, I do not yet see how GPUs/distributed computing can be supported seamlessly. Also NumPy is in many places Python specific or implemented in Python, so while a C backend/core library is a useful model/structure, I personally have doubts it should (or can) be a stated goal.
